src/Symfony/Component/JsonPath/JsonCrawler.php

```diff
@@ -562,8 +843,111 @@ private function evaluateRecursive(mixed $value): array
     private function compare(mixed $left, mixed $right, string $operator): bool
     {
         return match ($operator) {
-            '==' => $left === $right,
-            '!=' => $left !== $right,
+            '==' => $this->compareEquality($left, $right),
+            '!=' => !$this->compareEquality($left, $right),
+            '>', '>=', '<', '<=' => $this->compareOrdering($left, $right, $operator),
+            default => false,
+        };
+    }
+
+    private function compareEquality(mixed $left, mixed $right): bool
+    {
+        $leftIsNothing = $left instanceof Nothing;
+        $rightIsNothing = $right instanceof Nothing;
+
+        if (
+            $leftIsNothing && $rightIsNothing
+            || ($leftIsNothing && 0 === $right || 0 === $left && $rightIsNothing)
+            || $left === $right
+        ) {
+            return true;
+        }
+
+        if ($leftIsNothing || $rightIsNothing) {
+            return false;
+        }
+
+        // handle equality with potential scientific notation
+        if ((\is_int($left) || \is_float($left)) && (\is_int($right) || \is_float($right))) {
+            return (float) $left === (float) $right;
```

Comment:
I'm suspicious about the whole comparison logic (equality or ordering)
casting ints to float to compare them is risky
also, using `>` to compare strings will yield the wrong result for numeric strings
for number, the sign of `$a - $b` should be enough to decide
for strings, strncmp should be used I suppose
and for arrays, how do they compare? I didn't get the code around this aspect